File: leetcode/Python/find-and-replace-pattern.py
class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        def _replacePattern(s):
            um, sl, idx = dict(), [], 0
            for ch in s:
                if ch not in um:
                    um[ch]=str(idx)
                    idx+=1
                sl.append(um[ch])
            return ''.join(sl)
        sol, p= [], _replacePattern(pattern)
        for w in words:
            if p == _replacePattern(w):
                sol.append(w)
        return sol

File: leetcode/C++/find-and-replace-pattern.cpp
class Solution {
public:
    string transPattern(string &p){
        string s;
        unordered_map<char,char> um;
        for(int i=0,j=0; i<p.length(); ++i){
            if(um.find(p[i])==um.end()) um[p[i]]=(char)('a'+(++j));
            s.push_back(um[p[i]]);
        }
        return s;
    }
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> sol;
        pattern=transPattern(pattern);
        for(string &s : words){
            if(pattern==transPattern(s)) sol.push_back(s); 
        }
        return sol;
    }
};